### PR TITLE
k-Omega SST ABL capability

### PR DESCRIPTION
**Pull request by janiadcock**
**Pull-request type:**
- [ ] Bug fix 
- [ ] Documentation update
- [x] Feature enhancement

## Description of the pull-request

Added a RANS ABL option to the wall boundary conditions, which uses the existing Dirichlet no-slip condition for velocity and the existing high Reynolds number wall functions for k and omega. For BC calculations, uses roughness length as wall distance and a non-iterative equation to calculate friction velocity.

## Checklist

- [x] Builds successfully (*must test on at least one system/compiler combination*)
  - Operating systems 
    - [x] Linux
    - [ ] MacOS
  - Compilers 
    - [ ] GCC
    - [ ] LLVM/Clang
    - [x] Intel compilers
    - [ ] NVIDIA CUDA
- [x] Compiles without warnings
- [x] Passes all unit tests
- [x] Passes all regression tests
- [ ] Documentation builds without errors

*For new code updates*
- [x] Documentation updates - additions to user, theory, and verification manuals
- [ ] New unit tests providing coverage for new code, bug fixes etc.
- [x] New regression tests exercising the new code

### CODE REVIEW

File: src/SurfaceForceAndMomentAlgorithm.C [hunk 1]
```diff
@@ -54,7 +55,8 @@ SurfaceForceAndMomentAlgorithm::SurfaceForceAndMomentAlgorithm(
   const std::string &outputFileName,
   const int &frequency,
   const std::vector<double > &parameters,
-  const bool &useShifted)
+  const bool &useShifted,
+  const WallBoundaryConditionData &wallBCData)
```
Comment (sayerhs) [inline src/SurfaceForceAndMomentAlgorithm.C:59]: I feel that this capability already exists in SurfaceForceAndMomentWallFunctionAlgorithm. This class shouldn't be modified.

File: src/ngp_algorithms/SDRWallFuncAlg.C [hunk 2]
```diff
@@ -108,14 +113,20 @@ void SDRWallFuncAlg<BcAlgTraits>::execute()
...
 
-        DoubleType ypBip = 0.0;
-        for (int d=0; d < BcAlgTraits::nDim_; ++d) {
-          const DoubleType nj = v_area(ip, d) / aMag;
-          const DoubleType ej = 0.25 * (v_coord(nodeR, d) - v_coord(nodeL, d));
-          ypBip += nj * ej * nj * ej;
+        DoubleType ypBip;
+        if (RANSAblBcApproach_) {
+          // set ypBip to roughness height for wall function calculation
+          ypBip = z0_;
```
Comment (sayerhs) [inline src/ngp_algorithms/SDRWallFuncAlg.C:119]: Accessing instance member variables from device kernels will fail on GPUs.

File: src/ShearStressTransportEquationSystem.C [hunk 3]
```diff
@@ -191,8 +192,12 @@ void
 ShearStressTransportEquationSystem::register_wall_bc(
   stk::mesh::Part* part,
   const stk::topology& partTopo,
-  const WallBoundaryConditionData& /*wallBCData*/)
+  const WallBoundaryConditionData &wallBCData)
```
Comment (sayerhs) [inline src/ShearStressTransportEquationSystem.C:195]:
```suggestion
  const WallBoundaryConditionData& wallBCData)
```

File: src/Realm.C [hunk 4]
```diff
@@ -974,7 +974,13 @@ Realm::setup_post_processing_algorithms()
...
     // call through to the Eqsys
     if ( theType == "surface" ) {
-      equationSystems_.register_surface_pp_algorithm(theData);
+      for (size_t ibc = 0; ibc < boundaryConditions_.size(); ++ibc) {
+        BoundaryCondition& bc = *boundaryConditions_[ibc];
+        std::string name = physics_part_name(bc.targetName_);
+        if (bc.theBcType_ == WALL_BC) { 
+          equationSystems_.register_surface_pp_algorithm(theData, *reinterpret_cast<const WallBoundaryConditionData *>(&bc));
+        }
+      }
```
Comment (sayerhs) [inline src/Realm.C:983]: Recommend shifting this closer to where the SST wall function algorithms are registered.

File: src/NaluParsing.C [hunk 5]
```diff
@@ -781,6 +781,21 @@ namespace YAML
...
+    if (node["fixed_velocity"])
+    {
+      wallData.uFixed_ =
+          node["fixed_velocity"].as<double>();
+    }
+    if (node["fixed_height"])
+    {
+      wallData.zFixed_ =
+          node["fixed_height"].as<double>();
+    }
```
Comment (sayerhs) [inline src/NaluParsing.C:798]: If user activates RANS ABL boundary condition, we should make uRef and zRef parameters mandatory. 

```suggestion
    if (node["RANS_abl_bc"]) {
        const bool is_activated = node["RANS_abl_bc"].as<bool>();
        wallData.RANSAblBcApproach_ = is_activated;

        if (is_activated) {
            wallData.uFixed_ = node["fixed_velocity"].as<double>();
            wallData.zFixed_ = node["fixed_height"].as<double>();
        }
    }
```

File: docs/source/theory/boundaryConditions.rst [hunk 6]
```diff
@@ -794,6 +794,49 @@ surface as,
 
 .. _theory_open_bc:
 
+SST RANS Model for Atmospheric Boundary Layer
++++++++++++++++++++++++++++++++++++++++++++++++++++
+
+The following boundary conditions simulate the Atmospheric Boundary Layer, as described in Bautista, :cite:`Bautista:2011` and :cite:`Bautista:2015`. The Nalu-Wind SST RANS implementation matches the Monin-Obukhov profile when used with the model constants from Table-A I-1 (Boundreault, 2011) in :cite:`Bautista:2011` and the meshing method described  in :cite:`Bautista:2015`. The mesh described in :cite:`Bautista:2015` gives the Monin-Obukhov profile for roughness height 0.1. When the roughness height is decreased, the mesh must be refined near the wall. For example, for the :cite:`Bautista:2015` ABL test case using roughness height 0.001 instead of 0.1, the mesh size needs to be halved near the wall. 
+    
+The :math:`k` and :math:`\omega` boundary conditions are the same as in the Turbulent Kinetic Energy, :math:`k_{sgs}` LES model:Turbulent Kinetic Energy and Specific Dissipation SST High Reynolds Number Boundary conditions section:
```
[the remark this answers is not in the document]
Comment (janiadcock) [reply docs/source/theory/boundaryConditions.rst:802]: What do you mean by `:ref:`? I meant to `:cite:` the Bautista 2011 and 2015 papers. Are you referring to the " LES model:Turbulent Kinetic Energy and Specific Dissipation SST High Reynolds Number Boundary conditions" part? If so, yes, it would be great if I could do a reference there instead; how can I do this?

File: docs/source/theory/boundaryConditions.rst [hunk 7]
```diff
@@ -794,6 +794,49 @@ surface as,
...
+The :math:`k`, :math:`\omega`, and :math:`u` wall boundary conditions are set in the input file by specifying a wall boundary condition with ``RANS_abl_bc``. The input file must also specify a height and the velocity at that height with ``fixed_height`` and ``fixed_velocity``. 
+
+.. literalinclude:: ransAbl_wallUserData.i
+   :language: yaml
+
+This height, :math:`h`, and velocity, :math:`u_h`, could, for example, be the hub height of a wind turbine and the velocity measured at that height. 
+
+The input file should also include a momentum source term, ``momentum``. 
+
+.. literalinclude:: ransAbl_sourceTerm.i
```
[the remark this answers is not in the document]
Comment (janiadcock) [reply docs/source/theory/boundaryConditions.rst:823]: Okay. I was trying to be consistent with other input files in the documentation, which used the .i extension (e.g. /user/nalu_run/nalu.i)

File: docs/source/theory/boundaryConditions.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (sayerhs) [reply docs/source/theory/boundaryConditions.rst:802]: In [boundaryConditions.rst](https://github.com/Exawind/nalu-wind/blob/master/docs/source/theory/boundaryConditions.rst) add a link anchor 

```rst

.. _kwsst_high_re_bc:

Turbulent Kinetic Energy and Specific Dissipation SST High Reynolds Number Bound
ary conditions
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
~~~~~~~~~~~~~~
```

And then you can refer to it by doing 

```
:ref:`k-omega SST boundary conditions <kwsst_high_re_bc>
```

File: docs/source/theory/boundaryConditions.rst (hunk 7 above)
[the remark this answers is not in the document]
Comment (sayerhs) [reply docs/source/theory/boundaryConditions.rst:823]: Yeah, originally nalu used `.i` for input file suffix. We recently updated the regression test input files to `.yaml`. Seems like we didn't do a good job updating the docs. You can update them also.

File: src/Realm.C (hunk 4 above)
Comment (janiadcock) [reply src/Realm.C:983]: Where do you mean by that? Are you saying to move it outside of setup_post_processing_algorithms?

File: src/ngp_algorithms/SDRWallFuncAlg.C (hunk 2 above)
Comment (janiadcock) [reply src/ngp_algorithms/SDRWallFuncAlg.C:119]: Could you explain this issue more and give an example from the code of how to work around this issue? 

Is the problem that I can't use `RANSAblBcApproach_`, that I'm not allowed to create a variable such as `ypBip`, that I'm not allowed to have if/else statements, or something else? 

Thank you.

Comment (sayerhs) [reply src/ngp_algorithms/SDRWallFuncAlg.C:119]: You can't use `RANSAblBcApproach_` and `z0_` inside a `run_{entity, elem, face_elem}_algorithm` loop you'll have to create local copies within the execute function and capture that with the KOKKOS_LAMBDA. You can see how we have done this for other class instance variables

https://github.com/Exawind/nalu-wind/blob/5ce9f628676cdf7217330a3a2c832abfcdeb294f/src/ngp_algorithms/SDRWallFuncAlg.C#L77-L85

File: src/Realm.C (hunk 4 above)
Comment (sayerhs) [reply src/Realm.C:983]: Looking at what is being done in `SurfaceForceMoment*Alg`, that functionality is already available in `SurfaceForceMomentWallFuncAlg`.. if you switch to using that then you don't need to change the existing `register_surface_pp_algorithm` logic.

File: src/ngp_algorithms/WallFuncGeometryAlg.C [hunk 8]
```diff
@@ -25,7 +25,9 @@ namespace nalu {
 template<typename BcAlgTraits>
 WallFuncGeometryAlg<BcAlgTraits>::WallFuncGeometryAlg(
```
[the remark this answers is not in the document]
Comment (janiadcock) [reply src/ngp_algorithms/WallFuncGeometryAlg.C:26]: ShearStressTransportEquationSystem::register_wall_bc

File: src/SurfaceForceAndMomentAlgorithm.C (hunk 1 above)
Comment (janiadcock) [reply src/SurfaceForceAndMomentAlgorithm.C]: Could you explain the difference in the methods used to calculate uTau, tauWall (e.g. lambda vs. no lambda), and yplus in SurfaceForceAndMoment vs. SurfaceForceAndMomentWallFunction?
